doc/yaml_configuration.md

```diff
@@ -1200,3 +1200,65 @@ This field is convenient in setups that restrict access to GitHub, for instance
 
 
 Since 2.5.0
+
+## Hooks
+
+### GHC installation hooks (experimental)
+
+Stack's installation procedure can be fully customized by placing a shell script at
+`~/.stack/hooks/ghc-install.sh` and making it executable.
+
+The script **must** return an exit code of `0` and the standard output **must** be the
+absolute path to the ghc binary that was installed. Otherwise stack will ignore
+the hook and possibly fall back to its own installation procedure.
+
+Hooks are not run when `system-ghc: true`.
+
+When `install-ghc: false`, hooks are still run,
```

Comment:
This seems funny: if I don't want GHC to get installed, I don't want a hook up install it, either.

Using implicit config is also a little unnerving (unless stack already has a habit of doing so for hooks?). What if there were two config options, `ghc-install-override: <path/to/script>` and `ghc-install-use-fallback: <bool default True>`?

Reply:
`system-ghc: true` means that stack uses `PATH` to search for ghc, so running the hook makes no sense.

`install-ghc: true` already means that stack falls back to internal installation (e.g. if it can't find a suitable GHC in PATH).

So this perfectly aligns with current config semantics.

Reply:
Ok, we can ignore my suggested changes. :) I'll just try to describe my problems, which I don't think are addressed yet:

(NB: I'm not a regular contributor or maintainer to stack. Just sharing my opinions here.)

##### Problem 1: You're changing the semantics of the "install-ghc" flag.

Currently, `install-ghc: false` means "Do not install ghc". That is very clear and understandable.

It looks like you are modifying the meaning of that flag so that `install-ghc:false` means, "Don't install ghc... unless there is an install hook, in which case *do* install ghc using that hook."

I don't think that's a good change.

##### Problem 2 (Not a blocker): Implicit config

Changing behavior based on the presence and mode of particular files is not one of the good ideas to borrow from git. But I don't hold this position strongly. And I don't know how other users would feel about it.

Reply:
BTW you might be better off just creating a `use-ghcup` option. :D Unless you know of other people that would use the hook?

Reply:
> ##### Problem 1: You're changing the semantics of the "install-ghc" flag.
> 
> Currently, `install-ghc: false` means "Do not install ghc". That is very clear and understandable.
> 

Not really, since the hooks don't necessarily install GHC. The behavior of a hook is only determined by, well, the hook.


> 
> ##### Problem 2 (Not a blocker): Implicit config
> 
> Changing behavior based on the presence and mode of particular files is not one of the good ideas to borrow from git. But I don't hold this position strongly. And I don't know how other users would feel about it.

Pretty much all tools (ghc, stack, cabal, hls) change their behavior depending on presence and mode of files. So I'm not sure I agree.

Thanks for your input though.